### ISSUE TITLE
VideoPress: Block does not render the video on frontend after saving post right after completing video upload

### ISSUE THREAD

**Issue by dhasilva** [Automattic/jetpack#28727]
ref p1675369263214809/1675340425.645519-slack-C03TA48NSUX

Steps to reproduce:

- Go to the block editor
- Add a VideoPress block
- Upload a video
- Click on `Done` and _immediately after that_ click on `Save draft`
- Publish the post
- Visit the published page

The _iframe_ is not added, displaying the URL as text.
Changing the post and updating it does not solve it.

https://user-images.githubusercontent.com/8486249/216421066-58c1b47a-cbf7-4f9a-887e-a1c69791076b.mp4

**Comment by lhkowalski** [comment]
Replicated on both WoA and Jetpack sites. Had a hard time replicating on Simple, but seems like a matter of speed for doing the steps. Tried throttling the internet connection to delay the player loading but this does not help.

I was also able to simulate the non-fixable aspect, but after refreshing the block editor and adding more content to it, I could save the post again and have the video working.

I am jumping into some code details now, to better grasp how we request the HTML preview.

**Comment by lhkowalski** [comment]
I confirmed that this is also happening on v5.

**Comment by lhkowalski** [comment]
In the end, the problem is related to invalid oEmbed cache for the video on a given post (like pointed by @renatoagds on the original Slack thread).

I was able to take the block markup from one post to another and make it work on the new post even without the iframe markup on it.

Inspecting the `postmeta` table for the post with the problematic video, I spotted the invalid oembed cache entry:

<img width="474" alt="Screen Shot 2023-03-02 at 15 07 25" src="https://user-images.githubusercontent.com/6760046/222514290-6db80224-823a-4c6c-ac7a-fed9997cc5ac.png">

Cleaning this entry results on the block being rendered correctly on the post, and a new valid entry being created on the database. 

Given that saving it before the preview being loaded properly results on an invalid oEmbed cache entry, one quick shortcut to prevent the problem is locking the saving of the post while the video preview is not fully loaded, using:

```
wp.data.dispatch( 'core/editor' ).lockPostSaving( 'some-videopress-loading-lock-name' );
```

And once the loading is done, unlocking it with:

```
wp.data.dispatch( 'core/editor' ).unlockPostSaving( 'some-videopress-loading-lock-name' );
```

So the user can't save the post while the preview is not fully loaded.

I still need to figure out why the caching is invalid in this scenario, and how we could make the code resilient enought to discard the cache while rendering the block when the cache is invalid.

**Comment by lhkowalski** [comment]
I did some research regarding the oembed cache containing `{{unknown}}` values, and seems like it's not cleared automatically.

If we want the block to help on the recovey of wrong oembed cache, we may need to get all the oembed cache data for the given post, check for `unknown` values and remove them.

Given we can't really determine which oembed cache key is from a VideoPress embed, seems a bit intrusive if the plugin try to clear all of them, so probably trying to prevent it from happening in the first place is a better approach.

**Comment by lhkowalski** [comment]
I am exploring the lock option, have a working draft for it to check how it would behaves, planning to add a pull request for it today.

**Comment by lhkowalski** [comment]
Seems like the lock approach is not working the way I expected. The lock will not prevent the saving of a draft, just the saving of a published post. I still can replicate the bug after adding the lock...

**Comment by lhkowalski** [comment]
I digged a bit more deep into the oembed caching mechanism and found out that the `{{unknown}}` records are auto-fixed when a post containing them is saved again (references [1](https://core.trac.wordpress.org/ticket/14759#comment:49) and [2](https://core.trac.wordpress.org/ticket/14759#comment:53)).

The bug describes that changing the post and updating it won't fix the issue and I could replicate this behavior. I only got it auto-fixed (the way it should) when the `autosave` kicked in.

I was able to replicate the auto-fix by triggering the autosave with:

```
wp.data.dispatch('core/editor').autosave();
```

It's pretty reliable: when the bug is present (the `{{unknown}}` is on the database and the URL is showing on the post), triggering the autosave on the post will fix the problem 100% of the time.

Now I need to leverage this information to get a working solution for the bug. Maybe some autosaving right after the block is fully loaded for the first time or something in this line.

**Comment by lhkowalski** [comment]
I don't had much success on this new approach. The moments of the block lifecycle that I triggered the autosave function ensured a failed preview, meaning that this is exactly when I should not try to save the post :)

I still want to explore some alternative points for the saving, and also explore a backend approach using the `WP_Embed::cache_oembed` method. Keep 🚶 .
